crates/ironrdp-pdu/src/basic_output/bitmap/rdp6.rs

```diff
@@ -134,6 +134,31 @@ impl<'a> PduEncode for BitmapStream<'a> {
     }
 }
 
+impl<'a> BitmapStream<'a> {
+    pub fn encode_header(&self, dst: &mut WriteCursor<'_>) -> PduResult<()> {
```

Comment:
I think it is better to implement `PduEncode` trait here instead -- `BitmapStream` borrow their inner buffers instead of owning them, so there should be only minimal overhead when constructing `BitmapStream` with inner fields and actually encoding it via `ironrdp_pdu:;encode_cursor`. We will eventually migrate all pdu-related code to PduEncode/PduDecode`.

What do you think @mihneabuz ?

(Same comment for `BitmapUpdateData::encode_header`)

Reply:
The purpose for these methods is to reduce the amount of intermediary buffers and copying.
Take this function for example: https://github.com/Devolutions/IronRDP/pull/167/files/37954589ab655b8a570d8de3872f2b1f789b78ed#diff-abb6fe01531dc5aacd020389e7c572eab4257beae3708a7910a6b4577578b1abR143
In order to use `ironrdp_pdu::encode_cursor`, I would need to have another auxiliary buffer to hold the compressed data, borrow it into the `BitmapStream`, and then copy it into the destination buffer.

With this method, I can just encode the header of the `BitmapStream` and then write the compressed data directly into the destination buffer.

I agree that the code is a little weird/ugly, but this seemed like a good tradeoff for me.
Alternatively, I could try to split these Pdus into a separate header and data parts and implement `PduEncode/PduDecode` for them.

Do you have any other ideas? @pacmancoder @CBenoit

Reply:
Yes I see what you mean. Indeed, you could create another structure called `BitmapStreamHeader` that is taking care of encoding / decoding only the header part and `BitmapStream` itself could be refactored to reuse this struct in its `encode` and `decode` methods. This way you don’t need to reference an empty slice `&[]`.

Also, it seems that the `ColorPlanes` enum could be changed a bit to accommodate this refactor.

Currently it’s defined as:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorPlanes<'a> {
    Argb {
        data: &'a [u8],
    },
    AYCoCg {
        color_loss_level: u8,
        use_chroma_subsampling: bool,
        data: &'a [u8],
    },
}
```

But we could remove the common field `data` and rename it so we get `ColorPlaneDefinition`:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorPlaneDefinition {
    Argb,
    AYCoCg {
        color_loss_level: u8,
        use_chroma_subsampling: bool,
    },
}
```

And ultimately get something looking like:
```rust
pub struct BitmapStreamHeader {
    pub enable_rle_compression: bool,
    pub use_alpha: bool,
    pub color_plane_definition: ColorPlaneDefinition,
}

pub struct BitmapStream<'a> {
    pub header: BitmapStreamHeader,
    pub color_planes: &'a [u8], // <- this is where we moved the `data` field of the old enum `ColorPlanes`
}
```

What do you think? @pacmancoder @CBenoit

Reply:
Yes I see what you mean. Indeed, you could create another structure called `BitmapStreamHeader` that is taking care of encoding / decoding only the header part and `BitmapStream` itself could be refactored to reuse this struct in its `encode` and `decode` methods. This way you don’t need to reference an empty slice `&[]`.

Also, it seems that the `ColorPlanes` enum could be changed a bit to accommodate this refactor.

Currently it’s defined as:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorPlanes<'a> {
    Argb {
        data: &'a [u8],
    },
    AYCoCg {
        color_loss_level: u8,
        use_chroma_subsampling: bool,
        data: &'a [u8],
    },
}
```

But we could remove the common field `data` and rename it so we get `ColorPlaneDefinition`:
```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ColorPlaneDefinition {
    Argb,
    AYCoCg {
        color_loss_level: u8,
        use_chroma_subsampling: bool,
    },
}
```

And ultimately get something looking like:
```rust
pub struct BitmapStreamHeader {
    pub enable_rle_compression: bool,
    pub use_alpha: bool,
    pub color_plane_definition: ColorPlaneDefinition,
}

pub struct BitmapStream<'a> {
    pub header: BitmapStreamHeader,
    pub color_planes: &'a [u8], // <- this is where we moved the `data` field of the old enum `ColorPlanes`
}
```

What do you think? @pacmancoder @mihneabuz

Reply:
This is exactly what I had in mind!
As for [`BitmapUpdateData<'a>`](https://github.com/Devolutions/IronRDP/pull/167/files#diff-fa35c6d358b1347196b93a6d496da9869f8c68835d113b41357b1bd3d9614098R56), I think adding a header part is not ideal, as it would have to always be in sync with the len of the rectangles anyway. I can't think of a better way to do it without a separate `encode_header` method.

Reply:
@CBenoit Yep, I agree sounds perfect to me, I think it was an oversight on my behalf to keep `data` inside `ColorPlanes` in the first place 👍

Reply:
> This is exactly what I had in mind! As for [`BitmapUpdateData<'a>`](https://github.com/Devolutions/IronRDP/pull/167/files#diff-fa35c6d358b1347196b93a6d496da9869f8c68835d113b41357b1bd3d9614098R56), I think adding a header part is not ideal, as it would have to always be in sync with the len of the rectangles anyway. I can't think of a better way to do it without a separate `encode_header` method.

For BitmapUpdateData, let’s keep your approach for now. It’s not that bad, and if we come up with something better we can refactor at that time.